torch/autograd/__init__.py

```diff
@@ -561,6 +563,163 @@ def variable(*args, **kwargs):  # noqa: D103
     )
 
 
+def detect_grad_reference_cycles(parameters: Union[torch.Tensor, Sequence[torch.Tensor]], 
+                                  clear_grads: bool = False) -> list[torch.Tensor]:
+    r"""Detect and optionally clear reference cycles in gradient graphs.
+    
+    When using backward() with create_graph=True, a reference cycle is created
+    between parameters and their gradients, which can cause memory leaks. This
+    function helps detect these cycles and optionally clear them.
+    
+    Args:
+        parameters: A single tensor or sequence of tensors to check for cycles
+        clear_grads: If True, automatically set .grad to None for tensors with cycles
+        
+    Returns:
+        List of tensors that have reference cycles
+        
+    Example::
+    
+        >>> # After backward with create_graph=True
+        >>> loss.backward(create_graph=True)
+        >>> # Check for cycles
+        >>> leaked_params = torch.autograd.detect_grad_reference_cycles(model.parameters())
+        >>> if leaked_params:
+        ...     print(f"Warning: {len(leaked_params)} parameters have gradient cycles")
+        >>> # Or automatically clear them
+        >>> torch.autograd.detect_grad_reference_cycles(model.parameters(), clear_grads=True)
+        
+    .. warning::
+        This function uses Python's garbage collector to detect cycles, which may
+        have some overhead. Use it sparingly in performance-critical code.
+        
+    .. note::
+        A better approach is to use torch.autograd.grad() instead of backward()
+        when you need create_graph=True, as grad() doesn't create these cycles.
+    """
+    import gc
+    
+    if isinstance(parameters, torch.Tensor):
+        parameters = [parameters]
+    else:
+        parameters = list(parameters)
+    
+    leaked_tensors = []
+    
+    for param in parameters:
+        if not isinstance(param, torch.Tensor):
+            continue
+            
+        grad = param.grad
+        if grad is None:
+            continue
+            
+        # Check if the grad has a grad_fn (indicating create_graph=True was used)
+        if grad.grad_fn is not None:
+            # Check if this creates a reference cycle by seeing if the grad
+            # is reachable from the grad_fn's outputs
+            # For simplicity, we consider any grad with grad_fn as potentially cyclic
+            leaked_tensors.append(param)
+            
+            if clear_grads:
+                param.grad = None
+    
+    # Force garbage collection to clean up cycles if we cleared grads
+    if clear_grads and leaked_tensors:
+        gc.collect()
+    
+    return leaked_tensors
+
+
+def estimate_tensor_memory_overhead(tensors: Union[torch.Tensor, Sequence[torch.Tensor]],
```

Comment:
Please use typing.overload to indicate the two different return types based on whether detail is true or false